lib/parser/models.ex

```diff
@@ -0,0 +1,385 @@
+        updated_fields = Field.build_field_list(ex_name, fields)
+
+        %Exception{name: ex_name, fields: updated_fields}
+      end
+    end
+
+    defmodule Struct do
+      @moduledoc """
+      A thrift struct
+
+      The basic datastructure in Thrift, structs have aa name and a field list.
+      """
+      defstruct name: nil, fields: %{}
+      import Thrift.Parser.Conversions
+      alias Thrift.Parser.Models.Field
+
+      def new(name, fields) do
+        struct_name = atomify(name)
+        fields = Field.build_field_list(struct_name, fields)
+
+        %Struct{name: struct_name, fields: fields}
+      end
+    end
+
+    defmodule Union do
+      @moduledoc """
+      A thrift union
+
+      Unions can have one field set at a time.
+      """
+      defstruct name: nil, fields: %{}
+      import Thrift.Parser.Conversions
+      alias Thrift.Parser.Models.Field
+
+      def new(name, fields) do
+        name = atomify(name)
+        fields = Field.build_field_list(name, fields)
+        |> Enum.map(fn(field=%Field{}) ->
+          # According to thrift docs, unions have implicitly optional
+          # fields. See https://thrift.apache.org/docs/idl#union
+          %Field{field | required: false}
+        end)
+        %Union{name: name, fields: fields}
+      end
+
+    end
+
+    defmodule StructRef do
+      @moduledoc """
+      A reference to another struct.
+
+      While not a thrift type, this represents when a thrift type refers to
+      another.
+      """
+
+      defstruct referenced_type: nil
+      import Thrift.Parser.Conversions
+
+      def new(referenced_type) do
+        %StructRef{referenced_type: atomify(referenced_type)}
+      end
+    end
+
+    defmodule Service do
+      @moduledoc """
+      A thrift service
+
+      Services hold RPC functions and can extend other services.
+      """
+      defstruct name: nil, extends: nil, functions: []
+
+      import Thrift.Parser.Conversions
+
+      def new(name, functions, extends) do
+
+        %Service{name: atomify(name), extends: atomify(extends), functions: functions}
+      end
+    end
+
+    defmodule Function do
+      @moduledoc """
+      A thrift function
+
+      Functions are remote endpoints for thrift services. They contain an argument list, exceptions and return a typed object.
+      They can also be `oneway`, which means that thrift doesn't have to wait for
+      a reply from them.
+      """
+      defstruct oneway: false, return_type: :void, name: nil, params: [], exceptions: []
+      alias Thrift.Parser.Models.Field
+      import Thrift.Parser.Conversions
+
+      def new(oneway, return_type, name, params, exceptions) do
+        name = atomify(name)
+        params = Field.build_field_list(name, params)
+
+        %Function{
+          oneway: oneway,
+          return_type: return_type,
+          name: name,
+          params: params,
+          exceptions: exceptions
+        }
+      end
+    end
+
+    defmodule Program do
```

Comment:
I wonder if this would be better named "Schema" or "Grammar". What was the thinking behind "Program"?